File: src/train.py
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import StratifiedKFold, cross_val_score
from sklearn.metrics import accuracy_score


### Load Data
train = pd.read_csv("data/train.csv")
train = train.dropna()
test = pd.read_csv('data/test.csv')


### Transform Data
# Fill NaN values in Age column of the test set to median age 
# by Pclass and sex from train set
age_medians = train.groupby(['Pclass', 'Sex'])['Age'].median()

def fill_age(row):
    if pd.isna(row['Age']):
        return age_medians.loc[row['Pclass'], row['Sex']]
    else:
        return row['Age']

test['Age'] = test.apply(fill_age, axis=1)

# Convert Sex column into integer, 1 = male and 0 = female
train['Sex'] = (train['Sex'] == 'male').astype(int)
test['Sex'] = (test['Sex'] == 'male').astype(int)

# Convert Cabin into categorical column
train['Cabin'] = train['Cabin'].astype('category')
test['Cabin'] = test['Cabin'].astype('category')

# Convert Embarked into categorical column
train['Embarked'] = train['Embarked'].astype('category')
test['Embarked'] = test['Embarked'].astype('category')


### Train and Evaluate Model
# Initialize model
model = LogisticRegression()
k_folds = StratifiedKFold(n_splits = 5, shuffle=True, random_state=123)

# Select features to be trained on
train_x = train[['Sex', 'Age', 'SibSp', 'Parch', 'Pclass']]
train_y = train['Survived']
test_x = test[['Sex', 'Age', 'SibSp', 'Parch', 'Pclass']]

# Evaluate cross validation performance
scores = cross_val_score(model, train_x, train_y, cv=k_folds, scoring='accuracy')

print("Cross Validation Scores:", scores)
print("Average CV Score:", scores.mean())
print("Standard Deviation of CV Scores:", scores.std())

# Fit on full training data
model.fit(train_x, train_y)
pred = model.predict(train_x)

accuracy_score(train_y, pred)


### Make Predictions
test_pred = model.predict(test_x)

# Save to csv file
results = pd.DataFrame({
    'PassengerId': test['PassengerId'],
    'Survived': test_pred
})

results.to_csv

File: src/r/train.R
library(dplyr)

set.seed(123)

# ---- Load Data ----
train <- read.csv("data/train.csv")
train <- na.omit(train)
test <- read.csv("data/test.csv")

cat("---- Data Overview ----\n")
cat("Train Set:\n")
str(train)
cat("\nTest Set:\n")
str(test)


# ---- Transform Data ----
cat("\n\n---- Transforming Data ----\n")

# Fill missing Age values in test set with median Age by Pclass and Sex from train
age_medians <- train %>%
  group_by(Pclass, Sex) %>%
  summarize(median_age = median(Age, na.rm = TRUE))

test <- test %>%
  left_join(age_medians, by = c("Pclass", "Sex")) %>%
  mutate(Age = ifelse(is.na(Age), median_age, Age)) %>%
  select(-median_age)

cat("Filled missing Age values in test set using median Age by Pclass and Sex.\n")

# Convert Sex to numeric (1 = male, 0 = female)
train$Sex <- ifelse(train$Sex == "male", 1, 0)
test$Sex <- ifelse(test$Sex == "male", 1, 0)
cat("Converted 'Sex' column to binary variable (1 = male, 0 = female).\n")

# Drop Cabin Column
train <- train %>% select(-Cabin)
test <- test  %>% select(-Cabin)
cat("Dropped 'Cabin' column since it has many missing values.\n")

# Convert Embarked to categorical
train$Embarked <- as.factor(train$Embarked)
test$Embarked <- as.factor(test$Embarked)
cat("Converted 'Embarked' column to categorical variables.\n")


# ---- Train and Evaluate Model ----
cat("\n\n---- Training Model ----\n")

# Select features
features <- c("Sex", "Age", "SibSp", "Parch", "Pclass")
train_x <- train[, features]
train_y <- train$Survived

# Combine for glm
train_df <- cbind(train_x, Survived = train_y)

# Fit logistic regression
model <- glm(Survived ~ ., data = train_df, family = binomial)

cat("Model trained using glm with predictors: ", paste(features, collapse = ", "), "\n")


# ---- Evaluate on Training Set ----
train_pred_probs <- predict(model, newdata = train_x, type = "response")
train_pred <- ifelse(train_pred_probs > 0.5, 1, 0)
train_acc <- mean(train_pred == train_y)
cat(sprintf("Training Accuracy: %.4f\n", train_acc))


# ---- Make Predictions on Test Set ----
test_x <- test[, features]
test_pred_probs <- predict(model, newdata = test_x, type = "response")
test_pred <- ifelse(test_pred_probs > 0.5, 1, 0)

# Save predictions
results <- data.frame(
  PassengerId = test$PassengerId,
  Survived = test_pred
)
cat(nrow(results), "predictions made.\n")

cat("Preview of predictions:\n")
print(head(results))

write.csv(results, "test_predictions.csv", row.names = FALSE)
cat("Predictions saved to test_predictions.csv\n")

